pkg/kapp/diff/change_set_with_versioned_rs.go

```diff
@@ -76,6 +84,72 @@ func (d ChangeSetWithVersionedRs) Calculate() ([]Change, error) {
 	return allChanges, nil
 }
 
+func (d ChangeSetWithVersionedRs) checkAndUpdateResWithMaxDurationAnn(existingRsGrouped map[string][]ctlres.Resource,
+	existingRs versionedResources) error {
+
+	exNonVersionResMap := map[string]ctlres.Resource{}
+
+	for _, res := range existingRs.NonVersioned {
+		resKey := VersionedResource{res, nil}.UniqVersionedKey().String()
+		exNonVersionResMap[resKey] = res
+	}
+
+	for _, res := range d.newRs {
+		if val, found := res.Annotations()[maxDurationAnnKey]; found {
+			duration, err := time.ParseDuration(val)
+			if err != nil {
+				return fmt.Errorf("Unable to parse max-duration for resource %s: %s", res.Description(), err.Error())
+			}
+
+			resKey := VersionedResource{res, nil}.UniqVersionedKey().String()
+			exRes, found := exNonVersionResMap[resKey]
+			if !found {
+				if exResGroup, found := existingRsGrouped[resKey]; found {
+					exRes = exResGroup[len(exResGroup)-1]
+				}
+			}
+
+			if exRes != nil {
+				err = addLastRenewedTimeAnn(res, exRes, duration)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func addLastRenewedTimeAnn(res, exRes ctlres.Resource, duration time.Duration) error {
+	var (
+		lastRenewed, creationTime time.Time
+		err                       error
+	)
+
+	if val, found := exRes.Annotations()[lastRenewTimeAnnKey]; found {
+		lastRenewed, err = time.Parse(time.RFC3339, val)
+		if err != nil {
+			return fmt.Errorf("Unable to parse last-renewed-time for resource %s: %s", res.Description(), err.Error())
+		}
+	}
+
+	creationTime = exRes.CreatedAt()
+	if creationTime.Before(lastRenewed) {
+		creationTime = lastRenewed
+	}
+
+	if time.Now().Before(creationTime.Add(duration)) {
+		return nil
+	}
+	return ctlres.StringMapAppendMod{
+		ResourceMatcher: ctlres.AllMatcher{},
+		Path:            ctlres.NewPathFromStrings([]string{"metadata", "annotations"}),
+		KVs: map[string]string{
+			lastRenewTimeAnnKey: fmt.Sprintf("%v", time.Now().UTC().Format(time.RFC3339)),
+		},
+	}.Apply(res)
+}
+
 func (d ChangeSetWithVersionedRs) groupResources(rs []ctlres.Resource) map[string][]ctlres.Resource {
```

Comment:
We need to create a new struct in this file, something like `groupedVersionedResources{groupedResources map[string]ctlres.Resource}` it can have the function `newGroupedVersionedResources(...)`  which does the exact same thing that this function does and returns an object of type `groupedVersionedResource`

Reply:
We will refactor existing usage and can reuse this in our new file